futures-util/src/sink/with_flat_map.rs

```diff
@@ -60,74 +76,89 @@ where
         self.sink
     }
 
-    fn try_empty_stream(&mut self, cx: &mut task::Context) -> Poll<(), S::SinkError> {
-        if let Some(x) = self.buffer.take() {
-            match self.sink.poll_ready(cx)? {
-                Async::Ready(()) => self.sink.start_send(x)?,
-                Async::Pending => {
-                    self.buffer = Some(x);
-                    return Ok(Async::Pending);
+    fn try_empty_stream(self: PinMut<Self>, cx: &mut task::Context) -> Poll<Result<(), S::SinkError>> {
+        let WithFlatMap {
+            sink,
+            f: _,
+            stream,
+            buffer,
+            _phantom,
+        } = unsafe { PinMut::get_mut_unchecked(self) };
+        let mut sink = unsafe { PinMut::new_unchecked(sink) };
+        let mut stream = unsafe { PinMut::new_unchecked(stream) };
+
+
+        if let Some(x) = buffer.take() {
+            match try_poll!(sink.reborrow().poll_ready(cx)) {
+                Poll::Ready(()) => try_ready!(Poll::Ready(sink.reborrow().start_send(x))),
+                Poll::Pending => {
+                    *buffer = Some(x);
+                    return Poll::Pending;
                 }
-            }
+            };
         }
-        if let Some(mut stream) = self.stream.take() {
-            while let Some(x) = ready!(stream.poll_next(cx)) {
-                match self.sink.poll_ready(cx)? {
-                    Async::Ready(()) => self.sink.start_send(x)?,
-                    Async::Pending => {
-                        self.stream = Some(stream);
-                        self.buffer = Some(x);
-                        return Ok(Async::Pending);
+        if let Some(mut some_stream) = stream.reborrow().as_pin_mut() {
+            while let Some(x) = ready!(some_stream.reborrow().poll_next(cx)) {
+                let x = try_ready!(Poll::Ready(x));
+                match try_poll!(sink.reborrow().poll_ready(cx)) {
+                    Poll::Ready(()) => try_poll!(Poll::Ready(sink.reborrow().start_send(x))),
```

Comment:
`try_poll!(Poll::Ready...` is not so nice to read

Reply:
I don't disagree that it's awkward-- IMO the solution to this is to implement `Try` for `Poll` so that we can get reasonable `?` impls (so that `?` can be used on `Result<T, E>` to convert cleanly into `Poll<Result<T, E>>`) but that doesn't exist yet, so I'd prefer to leave as-is.